api/v1alpha1/nodemaintenance_types.go

```diff
@@ -20,27 +20,148 @@ import (
+	// +kubebuilder:validation:Optional
+	PodSelector string `json:"podSelector,omitempty"`
+
+	// TimeoutSecond specifies the length of time in seconds to wait before giving up drain, zero means infinite
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=300
+	// +kubebuilder:validation:Minimum:=0
+	TimeoutSecond uint32 `json:"timeoutSeconds,omitempty"`
+
+	// DeleteEmptyDir indicates if should continue even if there are pods using emptyDir
+	// (local data that will be deleted when the node is drained)
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=false
+	DeleteEmptyDir bool `json:"deleteEmptyDir,omitempty"`
+
+	// PodEvictionFilters specifies filters for pods that need to undergo eviction during drain.
+	// if specified. only pods that match PodEvictionFilters will be evicted during drain operation.
+	// if unspecified. all non-daemonset pods will be evicted.
+	// logical OR is performed between filter entires. logical AND is performed within different filters
+	// in a filter entry.
+	// +kubebuilder:validation:Optional
+	PodEvictionFilters []PodEvictionFiterEntry `json:"podEvictionFilters,omitempty"`
+}
+
+// PodEvictionFiterEntry defines filters for Pod evictions during drain operation
+type PodEvictionFiterEntry struct {
+	// ByResourceNameRegex filters pods by the name of the resources they consume using regex.
+	ByResourceNameRegex *string `json:"byResourceNameRegex,omitempty"`
 }
 
 // NodeMaintenanceStatus defines the observed state of NodeMaintenance
 type NodeMaintenanceStatus struct {
-	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
-	Bar string `json:"bar,omitempty"`
+	// Conditions represents observations of NodeMaintenance current state
+	// +kubebuilder:validation:Optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+
+	// Drain represents the drain status of the node
+	Drain *DrainStatus `json:"drain,omitempty"`
 }
 
-//+kubebuilder:object:root=true
-//+kubebuilder:subresource:status
+// DrainStatus represents the status of draining for the node
+type DrainStatus struct {
+	// TotalPods is the number of pods on the node at the time NodeMaintenance was Scheduled
+	TotalPods uint32 `json:"totalPods,omitempty"`
+
+	// EvictionPods is the total number of pods that need to be evicted at the time NodeMaintenance was scheduled
+	EvictionPods uint32 `json:"evictionPods,omitempty"`
+
+	// DrainProgress represents the draining progress as percentage
+	DrainProgress uint32 `json:"drainProgress,omitempty"`
+
+	// PendingPods is the list of pending namespaced named pods that need to be evicted
+	PendingPods []string `json:"pendingPods,omitempty"`
```

Comment:
For me it is not clear what is the meaning of this field. Is this list contains pods which are waiting for/in process of eviction?

Reply:
yes it is.  ive renamed to : "WaitingForEviction"